encoding/int_encode.go

```diff
@@ -1,30 +1,188 @@
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min number of bits to encode each actual value into the table.
+		enc.metaProto.BytesPerDictionaryValue = int64(math.Ceil(float64(bits.Len(uint(max))) / 8.0))
+		// Min number of bits required to encode table indices.
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(len(dict))))
+	default:
+		// Default to delta encoding.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DELTA
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(max - min)))
+	}
+
+	if err := proto.EncodeIntMeta(&enc.metaProto, &enc.buf, writer); err != nil {
+		return err
+	}
+
+	switch enc.metaProto.Encoding {
+	case encodingpb.EncodingType_DICTIONARY:
+		if err := enc.encodeDictionary(writer, enc.metaProto.BytesPerDictionaryValue, enc.metaProto.BitsPerEncodedValue, dict, valuesIt); err != nil {
+			return err
+		}
+	case encodingpb.EncodingType_DELTA:
+		if err := enc.encodeDelta(writer, enc.metaProto.BitsPerEncodedValue, valuesIt); err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 // Reset resets the encoder.
-func (enc *IntEnc) Reset() { panic("not implemented") }
+func (enc *IntEnc) Reset() {
+	enc.metaProto.Reset()
+	enc.dictionaryProto.Reset()
+}
+
+func (enc *IntEnc) encodeDelta(
+	writer io.Writer,
+	bitsPerEncodedValue int64,
+	valuesIt RewindableIntIterator,
+) error {
+	// Get first record and set as the start.
+	valuesIt.Next()
+	curr := valuesIt.Current()
+	enc.metaProto.DeltaStart = int64(curr)
+
+	// Encode the first value which is always a delta of 0.
+	bitWriter := bitstream.NewWriter(writer)
+	// Write an extra bit to encode the sign of the delta.
+	if err := bitWriter.WriteBits(uint64(0), int(bitsPerEncodedValue)+1); err != nil {
+		return err
+	}
+
+	var delta int
+	last := curr
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		delta = curr - last
+		if delta < 0 {
+			// Flip the sign.
+			delta *= -1
+			// Set the MSB if the sign is negative.
+			delta |= 1 << uint(bitsPerEncodedValue)
+		}
+		if err := bitWriter.WriteBits(uint64(delta), int(bitsPerEncodedValue)+1); err != nil {
+			return err
+		}
+		// Housekeeping.
+		last = curr
+	}
+
+	return valuesIt.Err()
+}
+
+func (enc *IntEnc) encodeDictionary(
+	writer io.Writer,
+	bytesPerDictionaryValue int64,
+	bitsPerEncodedValue int64,
+	dict map[int]int,
+	valuesIt RewindableIntIterator,
+) error {
+	sortedDict := make([]uint64, len(dict))
+	for value, idx := range dict {
+		sortedDict[idx] = uint64(value)
+	}
+
+	var start int
+	enc.dictionaryProto.Data = make([]byte, len(sortedDict)*int(bytesPerDictionaryValue))
+	for idx, value := range sortedDict {
+		endianness.PutUint64(enc.buf, value)
+		start = idx * int(bytesPerDictionaryValue)
+		copy(enc.dictionaryProto.Data[start:start+int(bytesPerDictionaryValue)], enc.buf[:bytesPerDictionaryValue])
+	}
+
+	if err := proto.EncodeIntDictionary(&enc.dictionaryProto, &enc.buf, writer); err != nil {
+		return err
+	}
+
+	bitWriter := bitstream.NewWriter(writer)
```

Comment:
Should probably make this a field of the encoder, and call `bitWriter.Reset(writer)` for every encode, and reuse the bitwriter here, so you don't need to create a new bit writer every time.

Reply:
👍